packages/components/src/sites-table/use-sites-table-filtering.tsx

```diff
@@ -1,77 +1,75 @@
 import { useFuzzySearch } from '@automattic/search';
 import { useI18n } from '@wordpress/react-i18n';
 import { useMemo } from 'react';
-// eslint-disable-next-line no-restricted-imports
-import type { SiteExcerptData } from 'calypso/data/sites/site-excerpt-types';
+import { SiteObjectWithStatus, getSiteStatus, siteStatuses } from './site-status';
 
-export type SiteStatus = 'all' | 'coming-soon' | 'private' | 'public';
+export const DEFAULT_SITE_STATUS_FILTER_VALUE = 'all';
+
+export const siteStatusFilterValues = [
+	DEFAULT_SITE_STATUS_FILTER_VALUE,
+	...siteStatuses,
+] as const;
+
+export type FilterableSiteStatuses = typeof siteStatusFilterValues[ number ];
 
 interface SitesTableFilterOptions {
-	status?: string;
+	status: FilterableSiteStatuses;
 	search?: string;
 }
 
 interface Status {
 	title: React.ReactChild;
-	name: SiteStatus;
+	name: FilterableSiteStatuses;
 	count: number;
 }
 
-interface UseSitesTableFilteringResult {
-	filteredSites: SiteExcerptData[];
+interface UseSitesTableFilteringResult< T > {
+	filteredSites: T[];
 	statuses: Status[];
 }
 
-export function useSitesTableFiltering(
-	allSites: SiteExcerptData[],
-	{ status = 'all', search }: SitesTableFilterOptions
-): UseSitesTableFilteringResult {
+type SiteObjectWithBasicInfo = SiteObjectWithStatus & {
+	URL: string;
+	name: string;
+	slug: string;
+};
+
+export function useSitesTableFiltering< T extends SiteObjectWithBasicInfo >(
+	allSites: T[],
+	{ status, search }: SitesTableFilterOptions
+): UseSitesTableFilteringResult< T > {
 	const { __ } = useI18n();
 
-	const [ statuses, filteredByStatus ] = useMemo( () => {
+	const [ statuses, groupedByStatus ] = useMemo( () => {
 		const statuses = [
 			{ name: 'all' as const, title: __( 'All Sites' ), count: 0 },
 			{ name: 'public' as const, title: __( 'Public' ), count: 0 },
 			{ name: 'private' as const, title: __( 'Private' ), count: 0 },
 			{ name: 'coming-soon' as const, title: __( 'Coming Soon' ), count: 0 },
 		];
 
-		const filteredByStatus = statuses.reduce(
-			( acc, { name } ) => ( { ...acc, [ name ]: filterSites( allSites, name ) } ),
-			{} as { [ name: string ]: SiteExcerptData[] }
+		const groupedByStatus = allSites.reduce< { [ K in Status[ 'name' ] ]: T[] } >(
+			( groups, site ) => {
+				const siteStatus = getSiteStatus( site );
+				groups[ siteStatus ].push( site );
+
+				return groups;
+			},
+			{ all: allSites, 'coming-soon': [], public: [], private: [] }
```

Comment:
Personally, I find this implementation of `allSites.reduce` to be very difficult to read (and the former version more so).

There isn't necessarily anything actionable from my comment — just wanted to state it.

Reply:
I'm not a huge fan either. I'd lean toward using `groupBy` but we're removing Lodash from the codebase...

Reply:
It's still possible to implement a `filterSites()` function using the new `getSiteStatus()` function.

```typescript
function filterSites< T >( sites: T[], filterType: FilterableSiteStatuses ): T[] {
	return sites.filter( ( site ) => getSiteStatus( site ) === filterType );
}
```

Which then means it's possible to do the group like we used to, but with the exhaustive-ness that @zaguiini mentioned.

```typescript
const groupedByStatus = statuses.reduce< Record< Status[ 'name' ], T[] > >(
	( acc, { name } ) => ( { ...acc, [ name ]: filterSites( allSites, name ) } ),
	{ all: allSites, 'coming-soon': [], public: [], private: [] }
);
```

Reply:
The thing is I'm not filtering -- I'm grouping. That's why I changed the iteratee from `statuses` to `allSites`, since we want to group the sites by status.

Reply:
Oh I see. My way has to loop over the `allSites` list 4 times, but yours is just once 👍